examples/pytorch/gat/main.py

```diff
@@ -92,58 +52,81 @@ def train_func(config: Dict):
         optimizer, named_parameters=model.named_parameters()
     )
 
-    address = "localhost:9999"
-    s = Server(address, args.data_dir, 0, len(args.partitions))
-    g = DistributedClient([address])
-    max_id = g.node_count(args.node_type) if args.max_id in [-1, None] else args.max_id
-    dataset = ray.data.range(max_id).repartition(max_id // args.batch_size)
-    pipe = dataset.window(blocks_per_window=4).repeat(args.num_epochs)
+    g = config["get_graph"]()
+    batch_size = 140
+    dataset = ray.data.read_text(f"{config['data_dir']}/train.nodes")
+    dataset = dataset.repartition(dataset.count() // batch_size)
+    pipe = dataset.window(blocks_per_window=4).repeat(config["num_epochs"])
 
     def transform_batch(idx: list) -> dict:
         return model.q.query_training(g, np.array(idx))
 
     pipe = pipe.map_batches(transform_batch)
 
-    for epoch in range(epochs_trained, args.num_epochs):
-        scores = []
-        labels = []
+    test_dataset = ray.data.read_text(f"{config['data_dir']}/test.nodes")
+    test_dataset = test_dataset.repartition(1)
+    test_dataset = test_dataset.map_batches(transform_batch)
+    test_dataset_iter = test_dataset.repeat(config["num_epochs"]).iter_epochs()
+
+    for epoch, epoch_pipe in enumerate(pipe.iter_epochs()):
+        model.train()
+        train_scores = []
+        train_labels = []
         losses = []
-        for step, batch in enumerate(dataset):
-            if step < steps_in_epoch_trained:
-                continue
+        for step, batch in enumerate(
+            epoch_pipe.iter_torch_batches(batch_size=batch_size)
+        ):
             loss, score, label = model(batch)
             optimizer.zero_grad()
             loss.backward()
             optimizer.step()
 
-            scores.append(score)
-            labels.append(label)
+            train_scores.append(score)
+            train_labels.append(label)
             losses.append(loss.item())
 
-        steps_in_epoch_trained = 0
-        if epoch % args.save_ckpt_by_epochs == 0:
-            save_checkpoint(model, logger, epoch, step, args)
+        model.eval()
+        batch = next(next(test_dataset_iter).iter_torch_batches(batch_size=1000))
+        loss, score, label = model(batch)
+        test_scores = [score]
+        test_labels = [label]
 
         session.report(
             {
-                "metric": model.compute_metric(scores, labels).item(),
+                "train_metric": model.compute_metric(train_scores, train_labels).item(),
+                "test_metric": model.compute_metric(test_scores, test_labels).item(),
                 "loss": np.mean(losses),
             },
         )
 
 
 def _main():
-    # setup default logging component.
     setup_default_logging_config(enable_telemetry=True)
-
     ray.init(num_cpus=4)
 
-    args = get_args(init_args)
+    data_dir = tempfile.TemporaryDirectory()
```

Comment:
why do we need to create a folder and can't use the `Cora` object directly?

Reply:
Ray re-initializes the GE many times so reloading a local client is slower and more memory consuming.

Reply:
We need to create a folder so we can tell where the sample files are

Reply:
Cora already has a `data_dir()` method..